pages/docs/couple-your-code/couple-your-code-gradient-data.md

```diff
@@ -0,0 +1,161 @@
+      bool          rowsFirst = false );
+```
+
+* `isDataGradientRequired` returns a boolean, indicates if the data corresponding to the ID `dataID` requires additional gradient data.
+* `writeScalarGradientData` writes gradient data corresponding to scalar data values, i.e., a vector containing the spatial derivative of a scalar quantity in each space dimension.
+* `ẁriteBlockScalarGradintData` writes multiple scalar gradient data at once (analogue to `writeBlockScalarData`).
+* `writeVectorGradientData` writes gradient data corresponding to vector-valued data, i.e, a Jacobian matrix. The matrix is passed as a 1D-array. The parameter `rowsFirst` allows to select whether the 1D-array contains the data column-wise or row-wise.
+* `ẁriteBlockVectorGradintData` writes multiple vector gradient data at once (analogue to `writeBlockVectorData`)
+
+Let's consider an example for writing block vector gradient data corresponding to the vector data `v0 = (v0x, v0y) , v1 = (v1x, v1y), ... , vn = (vnx, vny)` differentiated in spatial directions x and y.
+Per default, the values are passed as following:
+
+```cpp
+( v0x_dx, v0x_dy, v0y_dx, v0y_dy,
+  v1x_dx, v1x_dy, v1y_dx, v1y_dy,
+  ... ,
+  vnx_dx, vnx_dy, vny_dx, vny_dy  )  
+```
+
+When using the `rowsFirst` parameter, the following format is required:
+
+```cpp
+( v0x_dx, v0y_dx, v1x_dx, v1y_dx, ... , vnx_dx, vny_dx,
+  v0x_dy, v0y_dy, v1x_dy, v1y_dy, ... , vnx_dy, vny_dy  )
+```
+
+Let's add gradient data to our example code:
+
+```cpp
+turnOnSolver(); //e.g. setup and partition mesh
+
+precice::SolverInterface precice("FluidSolver","precice-config.xml",rank,size); // constructor
+
+// set mesh elements
+int dim = precice.getDimensions();
+int meshID = precice.getMeshID("FluidMesh");
+int vertexSize; // number of vertices at wet surface
+
+// determine vertexSize
+double* coords = new double[vertexSize*dim]; // coords of coupling vertices
+// determine coordinates
+int* vertexIDs = new int[vertexSize];
+precice.setMeshVertices(meshID, vertexSize, coords, vertexIDs);
+delete[] coords;
+
+// create data first
+int displID = precice.getDataID("Displacements", meshID);
+int stressID = precice.getDataID("Stress", meshID);
+
+double* Stress = new double[vertexSize * dim];
+double* displacements = new double[vertexSize * dim];
+
+// create gradient data
+double* stressGradient = new double[vertexSize * dim * dim]
+
+[...]
+precice_dt = precice.initialize();
+
+while (not simulationDone()){ // time loop
+  precice.readBlockVectorData(displID, vertexSize, vertexIDs, displacements);
+  setDisplacements(displacements);
+  [...]
+  solveTimeStep(dt);
+  computeStress(Stress);
+
+  precice.writeBlockVectorData(stressID, vertexSize, vertexIDs, Stress);
+
+  // write gradient data
+  if (isGradientDataRequired(dataID)){
+    computeStressGradient(stressGradient)
+    precice.writeBlockVectorGradientData(stressID, vertexSize, vertexIDs, stressGradient);
+  }
+
+  precice_dt = precice.advance(dt);
+}
+[...]
+```
```

Comment:
Let's shorten this some more:

```suggestion
```cpp
precice::SolverInterface precice("FluidSolver","precice-config.xml",rank,size); // constructor

int dim = precice.getDimensions();
int meshID = precice.getMeshID("FluidMesh");
[...] 
precice.setMeshVertices(meshID, vertexSize, coords, vertexIDs);

int stressID = precice.getDataID("Stress", meshID);
double* stress = new double[vertexSize * dim];

// create gradient data
double* stressGradient = new double[vertexSize * dim * dim]
[...]
precice_dt = precice.initialize();

while (not simulationDone()){ // time loop
  precice.readBlockVectorData(displID, vertexSize, vertexIDs, displacements);
  setDisplacements(displacements);
  [...]
  solveTimeStep(dt);
  computeStress(stress);

  precice.writeBlockVectorData(stressID, vertexSize, vertexIDs, stress);

  // write gradient data
  if (isGradientDataRequired(dataID)){
    computeStressGradient(stressGradient)
    precice.writeBlockVectorGradientData(stressID, vertexSize, vertexIDs, stressGradient);
  }

  precice_dt = precice.advance(dt);
}
[...]
```
```